### PR TITLE
Update Jetpack to webpack 5

### PR DESCRIPTION
**Pull request by gravityrail**
I wanted to include regular components, reducers, selectors, actions and action-types in component packages, but this becomes really messy when you try to import them because es6 imports don't support nesting. You end up with stuff like this:

```
import { State } from '@automattic/jetpack-plugins';
const { isPluginActive } = State;
```

However, since Node 12 there has been an "exports" field that allows you to map paths to subdirectories of the source tree. This allows us to do this instead:

```
import { isPluginActive } from '@automattic/jetpack-plugins/state';
```

Webpack 4 doesn’t support the package.json “exports” field, which is necessary if you want to do pathed exports like this:

```
"exports": {
  ".": ['index.js'],
  "./state": ['src/state/index.js'],
  "./action-types": ['src/state/action-types.js'],
}

// then in a package consumer:

import { SITE_RECONNECT } from '@automattic/jetpack-foo/action-types';
```

Hence, I upgraded Jetpack to Webpack 5. This PR includes the minimal set of changes I could find to make it work.

#### Changes proposed in this Pull Request:
* Upgrade Jetpack build to use Webpack 5

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:

`jetpack build plugins/jetpack` should result in a working build.

### CODE REVIEW

File: projects/plugins/jetpack/package.json [hunk 1]
```diff
@@ -180,14 +180,17 @@
 		"eslint": "7.31.0",
 		"fs-extra": "10.0.0",
 		"glob": "7.1.6",
+		"install": "0.13.0",
```
Comment (anomiex) [inline projects/plugins/jetpack/package.json:183]: What's this for?

File: projects/plugins/jetpack/package.json [hunk 2]
```diff
@@ -180,14 +180,17 @@
...
 		"jest": "27.0.4",
 		"jest-environment-jsdom": "^27.0.6",
 		"jetpack-js-test-runner": "workspace:*",
 		"lodash": "4.17.21",
 		"mocha": "8.2.1",
 		"mockery": "2.1.0",
 		"nock": "13.0.7",
+		"node-polyfill-webpack-plugin": "1.1.4",
 		"node-wp-i18n": "1.2.4",
+		"npm": "7.20.6",
```
Comment (anomiex) [inline projects/plugins/jetpack/package.json:193]: Even more, what's this for??

File: projects/plugins/jetpack/tools/webpack.config.extensions.js [hunk 3]
```diff
@@ -147,7 +147,9 @@ module.exports = [
 		resolve: {
 			...extensionsWebpackConfig.resolve,
 			// We want the compiled version, not the "calypso:src" sources.
-			mainFields: undefined,
+			mainFields: extensionsWebpackConfig.resolve.mainFields.filter(
+				entry => 'calypso:src' !== entry
+			),
```
Comment (anomiex) [inline projects/plugins/jetpack/tools/webpack.config.extensions.js:152]: I like this, good fix.

File: projects/plugins/jetpack/package.json (hunk 1 above)
Comment (gravityrail) [reply projects/plugins/jetpack/package.json:183]: That's weird, must have been put in there by `pnpm up ...` at some point. I didn't add it. I'll see if things break when I remove that and the other spurious dependencies.

File: projects/plugins/jetpack/package.json (hunk 2 above)
Comment (gravityrail) [reply projects/plugins/jetpack/package.json:193]: Yeah, weird. I didn't put it there... or at least, I don't _remember_ putting it there.

**Comment by anomiex** [comment]
Change itself looks good. I do hope you'll do a followup to update webpack and calypso-build for the other projects that use them. 🙂

Needs a master merge, though. Conflicting change in `pnpm-lock.yaml`.

The timeout in the wpcom tests is a little worrying, but it doesn't seem directly related to the changes in this PR.

Poking at it a bit in JN, things still seem to work.

Another matter for a followup, I see the build is outputting a lot of deprecation warnings.
* `[DEP_WEBPACK_EXTERNALS_FUNCTION_PARAMETERS] DeprecationWarning: The externals-function should be defined like ({context, request}, cb) => { ... }`
* `[DEP_WEBPACK_MODULE_ISSUER] DeprecationWarning: Module.issuer: Use new ModuleGraph API`
* `[DEP_WEBPACK_CHUNK_MODULES_ITERABLE] DeprecationWarning: Chunk.modulesIterable: Use new ChunkGraph API`
* `[DEP_WEBPACK_COMPILATION_ASSETS] DeprecationWarning: Compilation.assets will be frozen in future, all modifications are deprecated.`
* `[DEP_WEBPACK_DEPRECATION_ARRAY_TO_SET_PUSH] DeprecationWarning: chunk.files was changed from Array to Set (using Array method 'push' is deprecated)`
* `Using / for division is deprecated and will be removed in Dart Sass 2.0.0.`
Someone should probably go through those at some point, but I won't block this PR on that.

**Comment by gravityrail** [comment]
Just merged from master. I totally hear you about the deprecation warnings and other noise. I started down the path of fixing them but then decided that a smaller PR was better than a larger one, as long as they were just warnings and not breaking anything.

**Review by kraftbj — APPROVED** [review]

**Comment by gravityrail** [comment]
Deployed in r230592-wpcom

**Pull request merged**